### PR TITLE
Add intent order_id validation check in IPP capture endpoint

### LINKED ISSUE #4556: Harden intent validation during terminal payment capture
### Description

We should add more validation checks in the in-person payments flow to ensure that the payment intent has the necessary details to avoid order-intent mismatch errors.

### PR DESCRIPTION
**Pull request by anu-rock**
Fixes #4556

#### Changes proposed in this Pull Request

* Add validation check to ensure `order_id` set in payment intent's metadata matches the one received in the request.
* Log an error if validation fails.

##### REST API

* Create a new order in WooCommerce with "Cash on delivery" as the payment method.

* Create a payment intent in Stripe Dashboard at `https://dashboard.stripe.com/test/connect/accounts/<account_id>/payments/new`. (do not select the "capture funds later" option)

* Make a POST request to `/payments/orders/<order_id>/capture_terminal_payment`. Refer the [docs](https://github.com/woocommerce/woocommerce/wiki/Getting-started-with-the-REST-API) to set up authorization correctly.

* Verify that the payment fails with error code 409.

* On the intent details page `https://dashboard.stripe.com/test/connect/accounts/<account_id>/payments/<intent_id>`, scroll down to the "Metadata" section. Create a new key `order_id` with value from the order created in step 1.

* Make a POST request to `/payments/orders/<order_id>/capture_terminal_payment`.

* Verify that the payment succeeds this time.

##### Mobile app

* Set up the app for testing as per the guide at pdjVC1-9M-p2 

* Create a new order in WooCommerce with "Cash on delivery" as the payment method.

* Proceed to capture payment for the above order via the mobile app.

* Verify that the payment is captured successfully without errors.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests
- [x] Tested on mobile

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Review by mgascam — COMMENTED** [review]
I reviewed the code and everything looks good to me. I intend to do some manual testing also, will add my approval once it is done

**Review by dpaun1985 — APPROVED** [review]
👍 LGTM!
✅ Tested the API part and works as expected!

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php [hunk 1]
```diff
@@ -297,6 +316,41 @@ public function test_capture_terminal_payment_intent_non_capturable() {
...
+				'payment_intent_id' => $this->mock_intent_id,
+			]
+		);
+
+		$response = $this->controller->capture_terminal_payment( $request );
+
+		$this->assertInstanceOf( 'WP_Error', $response );
+		$data = $response->get_error_data();
+		$this->assertArrayHasKey( 'status', $data );
+		$this->assertEquals( 409, $data['status'] );
```
Comment (kalessil) [inline tests/unit/admin/test-class-wc-rest-payments-orders-controller.php]: Can we use `assertSame` here?

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 2]
```diff
@@ -124,7 +124,14 @@ public function capture_terminal_payment( $request ) {
 			}
 
 			// Do not process intents that can't be captured.
-			$intent = $this->api_client->get_intent( $intent_id );
+			$intent                   = $this->api_client->get_intent( $intent_id );
+			$intent_metadata          = is_array( $intent->get_metadata() ) ? $intent->get_metadata() : [];
+			$intent_meta_order_id_raw = $intent_metadata['order_id'] ?? '';
+			$intent_meta_order_id     = is_numeric( $intent_meta_order_id_raw ) ? intval( $intent_meta_order_id_raw ) : 0;
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-orders-controller.php:130]: As `$intent_metadata['order_id'] ?? ''` will tolerate non-array data structures, we can simplify this a little bit:

```php
$intent_meta_order_id = $intent->get_metadata()['order_id'] ?? '';
$intent_meta_order_id = is_numeric( $intent_meta_order_id ) ? (int) $intent_meta_order_id : 0;
```

**Review by kalessil — APPROVED** [review]
LGTM! I left a couple suggestions with possible tweaks.

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php (hunk 1 above)
Comment (anu-rock) [reply tests/unit/admin/test-class-wc-rest-payments-orders-controller.php]: Fixed in f528fed.

File: includes/admin/class-wc-rest-payments-orders-controller.php (hunk 2 above)
Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-orders-controller.php:130]: Fixed in 2297745.

Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-orders-controller.php:130]: I had to revert this change because of the error mentioned in 18462d9.

Comment (kalessil) [reply includes/admin/class-wc-rest-payments-orders-controller.php:130]: Gotcha, thanks for trying it out!

**Pull request merged**